modules/dag-l1/src/main/scala/org/tessellation/dag/l1/infrastructure/block/BlockService.scala

```diff
@@ -0,0 +1,173 @@
+
+      def handleTipsUpdate(hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        for {
+          allBlocks <- blocks.toMap
+          allTips = allBlocks.collect { case (hash, ab @ AcceptedBlock(_, _, _, true)) => hash -> ab }
+          parentHashToTip = hashedBlock.signed.value.parent.map(ref => ref.hash -> allTips.get(ref.hash)).toList.toMap
+          notTipParents = parentHashToTip.collect { case (hash, None) => hash }.toList
+          tipParents = parentHashToTip.collect { case (hash, Some(accepted)) => hash -> accepted }
+          tipParentsSize = tipParents.size
+          tipParentsToRemove = tipParents.filter {
+            case (_, block) =>
+              allTips.size >= tipsConfig.minimumTipsCount + tipParentsSize &&
+                (
+                  allTips.size > tipsConfig.maximumTipsCount + tipParentsSize ||
+                    block.usages.size >= tipsConfig.maximumTipUsages - 1
+                )
+          }
+          tipParentsToRemain = tipParents -- tipParentsToRemove.keySet
+          remainingTips = allTips -- tipParentsToRemove.keySet
+          minTipHeight = remainingTips.values.map(_.block.signed.value.height.coerce).minOption.getOrElse(0L)
+          addAsTipCondition = remainingTips.size < tipsConfig.maximumTipsCount && minTipHeight < hashedBlock.signed.value.height.coerce
+          _ <- updateNoTipParents(notTipParents, hashedBlock)
+          _ <- updateTipParentsToRemove(tipParentsToRemove.keySet.toList, hashedBlock)
+          _ <- updateTipParentsToRemain(tipParentsToRemain.keySet.toList, hashedBlock)
+          _ <- tryAddAsATip(addAsTipCondition, hashedBlock)
+        } yield ()
+
+      def pullTips(tipsCount: PosInt): F[Option[PulledTips]] =
+        for {
+          allBlocks <- blocks.toMap
+          allTips = allBlocks.collect { case (_, AcceptedBlock(block, _, _, true)) => block }.toList
+            .sortBy(_.signed.value.height.coerce)
+          pulledTips = allTips match {
+            case tips if tips.size >= tipsCount =>
+              NonEmptyList
+                .fromList(tips.take(tipsCount))
+                .map(parents => block.PulledTips(parents.map(p => BlockReference(p.proofsHash, p.signed.value.height))))
+            case _ => None
+          }
+        } yield pulledTips
+
+      private def isBlockAccepted(blockReference: BlockReference): F[Boolean] =
+        blocks(blockReference.hash).get.map(_.exists(_.isInstanceOf[AcceptedBlock]))
+
+      private def updateNoTipParents(parents: List[ProofsHash], hashedBlock: Hashed[DAGBlock]): F[Unit] =
+        parents.traverse { hash =>
+          blocks(hash)
+            .modify[Either[TipsUpdateError, Unit]] {
+              case Some(accepted @ AcceptedBlock(_, _, _, false)) =>
+                (accepted.addUsage(hashedBlock).some, ().asRight)
+              case unexpected =>
+                (unexpected, NotTipAcceptedBlockNotFound(hash, unexpected).asLeft)
```

Comment:
`not tip accepted block not found` - sorry, but I don't understand it :( 
tip is not accepted? or block is not found? or what? :)

Reply:
It means: Accepted block that's not a tip wasn't found. The name starting with `Not` maybe misleading but if I change it to `NoTipAcceptedBlockNotFound` I think it works.

Reply:
So I'd propose a general type like `TipNotFound` or `BlockNotFound` with specific details inside, like a part of round, etc.
```
case class BlockNotFound( shouldBeTip: yes | no, state: Any | Accepted | ..., hash: ...  ) ...
```

Or just more verbose names like `NoTipBlockNotFound(when: Accepted)` or `TipBlockNotFound(when: Unset)` or sth like that 🤔

Reply:
If that's ok I would still like to go with the current approach. If we write the rest of the implementation correctly we shouldn't get many more errors like this added to the service. And I think it is easier to pinpoint where the error happened with a concrete error than with a more generic one. I like the one-to-one correspondence between an error and a place in the code.